test/e2e/network/ingressclass.go

```diff
@@ -50,13 +50,35 @@ var _ = common.SIGDescribe("IngressClass [Feature:Ingress]", func() {
 		framework.ExpectNoError(err)
 		defer deleteIngressClass(cs, ingressClass1.Name)
 
-		ingress, err := createBasicIngress(cs, f.Namespace.Name)
-		framework.ExpectNoError(err)
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		lastFailure := ""
+
+		// the admission controller may take a few seconds to observe the ingress classes
+		if err := wait.Poll(time.Second, time.Minute, func() (bool, error) {
+			lastFailure = ""
+
+			ingress, err := createBasicIngress(cs, f.Namespace.Name)
+			if err != nil {
+				lastFailure = err.Error()
+				return false, err
+			}
+			defer func() {
+				err := cs.NetworkingV1().Ingresses(ingress.Namespace).Delete(ctx, ingress.Name, metav1.DeleteOptions{})
+				framework.Logf("%v", err)
+			}()
+
+			if ingress.Spec.IngressClassName == nil {
+				lastFailure = "Expected IngressClassName to be set by Admission Controller"
+				return false, nil
+			} else if *ingress.Spec.IngressClassName != ingressClass1.Name {
+				lastFailure = fmt.Sprintf("Expected IngressClassName to be %s, got %s", ingressClass1.Name, *ingress.Spec.IngressClassName)
+				return false, nil
+			}
+			return true, nil
 
-		if ingress.Spec.IngressClassName == nil {
-			framework.Failf("Expected IngressClassName to be set by Admission Controller")
-		} else if *ingress.Spec.IngressClassName != ingressClass1.Name {
-			framework.Failf("Expected IngressClassName to be %s, got %s", ingressClass1.Name, *ingress.Spec.IngressClassName)
+		}); err != nil || len(lastFailure) != 0 {
```

Comment:
when does the loop exits with `err == nil` and `len(lastFailure) > 0`

Reply:
`lastFailure` is cleared at the start of every attempt. `lastFailure` is only set when errors happen, so if you reach `return true, nil`, the error will be nil and the lastFailure will be empty, right?

Reply:
yeah, but when err != is nil `lastFailure`is never empty, I may be wrong but I think can we check `err !=nil`only

Reply:
Look at L73 and L76 ?

Reply:
yeah, but if it `return false,nil` it retries and is cleared, and if it keeps returning false it eventually times out and returns an error
https://github.com/kubernetes/kubernetes/blob/57eb5d631ccd615cd161b6da36afc759af004b93/staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go#L594

if `returns false, err !=nil` is the same as timeout `err !=nil `
if `returns true, nil` then `lastFailure = ""` 

am I again missing something obvious like in https://github.com/kubernetes/kubernetes/pull/112955/files#r1029922666 ?

Reply:
o h, yes, I see.  That's the condition func.  You are correct.  The 2nd clause is pointless, unless I am missing something - @deads2k  ?

Reply:
> o h, yes, I see. That's the condition func. You are correct. The 2nd clause is pointless, unless I am missing something - @deads2k ?

accepted and updated.